### PR TITLE
fix issue #71 (new ccall api)

### PR DESCRIPTION
**Pull request by skariel**

### CODE REVIEW

File: src/ZMQ.jl [hunk 1]
```diff
@@ -404,7 +404,7 @@ type Message <: AbstractArray{Uint8,1}
     # Create a message with a given AbstractString or Array as a buffer (for send)
     # (note: now "owns" the buffer ... the Array must not be resized,
     #        or even written to after the message is sent!)
-    Message(m::ByteString) = Message(m, convert(Ptr{Uint8}, m), sizeof(m))
+    Message(m::ByteString) = Message(m, Base.unsafe_convert(Ptr{Uint8}, pointer(m.data)), sizeof(m))
```
Comment (vtjnash) [inline src/ZMQ.jl]: `pointer(m.data)` should work with just `m`

**Comment by jiahao** [comment]
I think for backwards compatibility you'd want to do something like

```jl
if VERSION <= v"0.4.0-dev+3703"
    old code
else
   new code
end
```

The alternative is to figure out a general solution that can be placed in `Compat.jl`, but this might be impossible.

File: src/ZMQ.jl [hunk 2]
```diff
@@ -501,8 +501,8 @@ function send(socket::Socket, zmsg::Message, flag=int32(0))
     if (get_events(socket) & POLLOUT) == 0
         wait(socket; writable = true)
     end
-    rc = ccall((:zmq_msg_send, zmq), Cint, (Ptr{Void}, Ptr{Message}, Cint),
-                &zmsg, socket.data, flag)
+    rc = ccall((:zmq_msg_send, zmq), Cint, (Ref{Message}, Ptr{Message}, Cint),
+                zmsg, socket.data, flag)
```
Comment (vtjnash) [inline src/ZMQ.jl]: this wasn't supposed to be deprecated (yet). although i see the argument types were reversed (http://api.zeromq.org/3-3:zmq-msg-send), passing `&arg` to `Ptr{Void}` was supposed to keep working

**Comment by skariel** [comment]
now passing tests on all Julia versions

**Comment by stevengj** [comment]
Thanks.

In the long run, we really need to figure out some `Compat` syntax.  Maybe `@compat Ref{T}` could turn into `Ptr{T}` or something?

**Comment by vtjnash** [comment]
> +    if VERSION <= v"0.4.0-dev+3703"
+        rc = ccall((:zmq_msg_send, zmq), Cint, (Ptr{Void}, Ptr{Message}, Cint),
+                    &zmsg, socket.data, flag)
+    else
+        rc = ccall((:zmq_msg_send, zmq), Cint, (Ref{Message}, Ptr{Message}, Cint),
+                    zmsg, socket.data, flag)
+    end  

this could also have been fixed by correcting the declared calling convention of the 0.3 version to match the header file. (although passing `&value` to `Ptr{Void}` was supposed to still work too, for now)

the `&` syntax is not deprecated yet in ccall, in part because there's no way to achieve quite the same effect in 0.3.

we should add an Compat.jl definition: `eval(Base, :(const unsafe_convert = convert))`

**Pull request merged**